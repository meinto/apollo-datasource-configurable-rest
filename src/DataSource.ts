import { RESTDataSource } from 'apollo-datasource-rest'
import { URLSearchParams, RequestInit, BodyInit } from 'apollo-server-env'
import { Headers as RequestHeaders } from 'apollo-env'

export type PrivateConfig<TArgs, TParams, THeaders, TBody> = {
  url: string,
  defaultArgs: Partial<TArgs>,
  params: TParams,
  headers: THeaders,
  body: TBody,
  cacheTime: number,
}

export type RequestInitOptions = {
  cacheTime?: number
}

export type Body = BodyInit | object

export class ConfigurableRESTDataSource<
  TArgs = Record<string, string>,
  TParams = Record<string, string>, 
  THeaders = Record<string, string>, 
  TBody = Body, 
> extends RESTDataSource {
  private config: PrivateConfig<TArgs, TParams, THeaders, TBody> = {
    url: '',
    defaultArgs: {} as TArgs,
    params: {} as TParams,
    headers: {} as THeaders,
    body: {} as TBody,
    cacheTime: 5,
  }

  protected set url(url: string) {
    this.config.url = url
  }
  protected set defaultArgs(args: Partial<TArgs>) {
    this.config.defaultArgs = args
  }
  protected set params(params: TParams) {
    this.config.params = params
  }
  protected set headers(headers: THeaders) {
    this.config.headers = headers
  }
  protected set body(body: TBody) {
    this.config.body = body
  }
  protected set cacheTime(cacheTime: number) {
    this.config.cacheTime = cacheTime
  }

  protected get url(): string {
    return this.config.url
  }
  protected get defaultArgs(): Partial<TArgs> {
    return this.config.defaultArgs
  }
  protected get params(): TParams {
    return this.config.params
  }
  protected get headers(): THeaders {
    return this.config.headers
  }
  protected get body(): TBody {
    return this.config.body
  }
  protected get cacheTime() {
    return this.config.cacheTime
  }

  public async configuredGET<TResult = any>(args: Partial<TArgs>, options?: RequestInitOptions): Promise<TResult> {
    return this.get<TResult>(
      this.replaceArgsInString(this.url, this.getArgs(args)),
      this.convertToURLSearchParams(this.params, this.getArgs(args)),
      this.getRequestInit(this.getArgs(args), options),
    )
  }

  public async configuredDELETE<TResult = any>(args: Partial<TArgs>, options?: RequestInitOptions): Promise<TResult> {
    return this.delete<TResult>(
      this.replaceArgsInString(this.url, this.getArgs(args)),
      this.convertToURLSearchParams(this.params, this.getArgs(args)),
      this.getRequestInit(this.getArgs(args), options),
    )
  }

  public async configuredPOST<TResult = any>(args: Partial<TArgs>, body?: Body): Promise<TResult> {
    const url = new URL(this.replaceArgsInString(this.url, this.getArgs(args)))
    url.search = this.convertToURLSearchParams(this.params, this.getArgs(args)).toString()
    return this.post<TResult>(
      url.toString(),
      this.convertToBody((body || this.body) as Body, this.getArgs(args)),
      this.getRequestInit(this.getArgs(args)),
    )
  }

  public async configuredPUT<TResult = any>(args: Partial<TArgs>, body?: Body): Promise<TResult> {
    const url = new URL(this.replaceArgsInString(this.url, this.getArgs(args)))
    url.search = this.convertToURLSearchParams(this.params, this.getArgs(args)).toString()
    return this.put<TResult>(
      url.toString(),
      this.convertToBody((body || this.body) as Body, this.getArgs(args)),
      this.getRequestInit(this.getArgs(args)),
    )
  }

  public async configuredPATCH<TResult = any>(args: Partial<TArgs>, body?: Body): Promise<TResult> {
    const url = new URL(this.replaceArgsInString(this.url, this.getArgs(args)))
    url.search = this.convertToURLSearchParams(this.params, this.getArgs(args)).toString()
    return this.patch<TResult>(
      url.toString(),
      this.convertToBody((body || this.body) as Body, this.getArgs(args)),
      this.getRequestInit(this.getArgs(args)),
    )
  }

  private getArgs(args: Partial<TArgs>): TArgs {
    return {
      ...this.defaultArgs,
      ...args,
    } as TArgs
  }

  private replaceArgsInString(str: string, args: TArgs): string {
    let replaced = str
    for (const key in args) {
      const regexp = new RegExp(`\\$${key}`, 'g')
      replaced = replaced.replace(regexp, this.convertToString(args[key]))
    }
    return replaced
  } 

  private replaceArgsInObject<TObject = Record<string, string>>(obj: TObject, args: TArgs): TObject {
    let str = this.replaceArgsInString(JSON.stringify(obj), args)
    return JSON.parse(str)
  }

  private convertToURLSearchParams(params: TParams, args: TArgs): URLSearchParams {
    const paramsWithReplacedArgs = this.replaceArgsInObject<TParams>(params, args)
    const urlParams = new URLSearchParams()
    for (const key in paramsWithReplacedArgs) {
      const convertedValue = this.convertToString(paramsWithReplacedArgs[key])
      if (convertedValue.indexOf('$') === -1) urlParams.append(key, convertedValue)
    }
    return urlParams
  }

  private convertToHeaders(headers: THeaders, args: TArgs): RequestHeaders {
    const headersWithReplacedArgs = this.replaceArgsInObject<THeaders>(headers, args)
    const requestHeaders = new RequestHeaders()
    for (const key in headersWithReplacedArgs) {
      const convertedValue = this.convertToString(headersWithReplacedArgs[key])
      if (convertedValue.indexOf('$') === -1) requestHeaders.append(key, convertedValue)
    }
    return requestHeaders
  }

  private convertToBody(body: Body, args: TArgs): Body {
    if (body instanceof String) {
      return this.replaceArgsInString(body as string, args)
    }
    if (body instanceof Object) {
      return this.replaceArgsInObject(body, args)
    }
    return body
  }

  convertToString(val: string | Object): string {
    let converted = `${val}`
    if (val instanceof Object) {
      converted = JSON.stringify(val)
    }
    return converted
  }

  private getRequestInit(args: TArgs, options?: RequestInitOptions): RequestInit {
    const ttl = (options && options.cacheTime) || this.cacheTime
    return {
      cacheOptions: { ttl },
      headers: this.convertToHeaders(this.headers, args)
    }
  }
}