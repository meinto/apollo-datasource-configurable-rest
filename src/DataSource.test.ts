import { Headers as RequestHeaders } from 'apollo-env'
import { URLSearchParams } from 'apollo-server-env'
import { ConfigurableRESTDataSource } from './DataSource'

jest.mock('apollo-datasource-rest', () => ({
  RESTDataSource: class {
    public get = async (...args: any) => args
    public post = async (...args: any) => args
    public put = async (...args: any) => args
    public delete = async (...args: any) => args
    public patch = async (...args: any) => args
  }
}))

type Args = {
  arg1: string,
  arg2: string,
  default: string,
}

type Params = {
  param1: string,
  param2: number,
  param3?: {
    param31: string,
    param32: number,
  }
}

type Headers = {
  header1: string,
  header2: string,
  header3: {
    headerField31: string
    headerField32: string
  }
}

type Body = {
  bodyField1: string,
  bodyField2: string,
  bodyField3: {
    bodyField31: string,
    bodyField32: string,
  },
}

const defaultArg = 'default-argument'
class TestDataSource extends ConfigurableRESTDataSource<Args, Params, Headers, Body> {
  url = "https://test.url/$arg1/$arg2/$default"
  defaultArgs = {
    default: defaultArg,
  }
  params = {
    param1: "$arg1",
    param2: 22,
    param3: {
      param31: "$arg2",
      param32: 33,
    }
  }
  headers = {
    header1: "$arg1",
    header2: "$default",
    header3: {
      headerField31: "arg2: $arg2",
      headerField32: "headerField32"
    }
  }
  body = {
    bodyField1: "$arg2",
    bodyField2: "$default",
    bodyField3: {
      bodyField31: "arg2: $arg2",
      bodyField32:"arg2: $arg2",
    }
  }
  cacheTime = 5000
}

class TestDataSourceWithoutSearchParams extends ConfigurableRESTDataSource<Args, Params, Headers, Body> {
  url = "https://test.url/$arg1/$arg2/$default"
  defaultArgs = {
    default: defaultArg,
  }
  headers = {
    header1: "$arg1",
    header2: "$default",
    header3: {
      headerField31: "arg2: $arg2",
      headerField32: "headerField32"
    }
  }
  body = {
    bodyField1: "$arg2",
    bodyField2: "$default",
    bodyField3: {
      bodyField31: "arg2: $arg2",
      bodyField32:"arg2: $arg2",
    }
  }
  cacheTime = 5000
}

describe('ConfigurableRESTDataSource', () => {

  const args = {
    arg1: "mock-arg-1",
    arg2: "mock-arg-2",
  }

  const expectedParams = () => {
    const params = new URLSearchParams()
    params.append('param1', `${args.arg1}`)
    params.append('param2', '22')
    params.append('param3', JSON.stringify({
      param31: `${args.arg2}`,
      param32: 33,
    }))
    return params
  }
  const expectedHeaders = () => {
    const headers = new RequestHeaders()
    headers.append('header1', `${args.arg1}`)
    headers.append('header2', defaultArg)
    headers.append('header3', JSON.stringify({
      headerField31: `arg2: ${args.arg2}`,
      headerField32: "headerField32"
    }))
    return headers
  }
  const expectedBody = () => ({
    bodyField1: `${args.arg2}`,
    bodyField2: defaultArg,
    bodyField3: {
      bodyField31: `arg2: ${args.arg2}`,
      bodyField32: `arg2: ${args.arg2}`,
    }
  })
  const expectedRequestInit = (headers = expectedHeaders()) => {
    return {
      cacheOptions: { ttl: 5000 },
      headers,
    }
  }

  describe('datasource with search params', () => {
    const datasource = new TestDataSource()

    it('removes search params and headers when the corresponding arguments are not available', async () => {
      const clonedArgs = {...args}
      delete clonedArgs.arg1
      const result = await datasource.configuredGET(clonedArgs)

      const params = new URLSearchParams()
      params.append('param2', '22')
      params.append('param3', JSON.stringify({
        param31: `${args.arg2}`,
        param32: 33,
      }))
      
      const headers = new RequestHeaders()
      headers.append('header2', defaultArg)
      headers.append('header3', JSON.stringify({
        headerField31: `arg2: ${args.arg2}`,
        headerField32: "headerField32"
      }))

      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(params)
      expect(result[2]).toEqual(expectedRequestInit(headers))
    })

    it('calls configuredGET with expected args', async () => {
      const result = await datasource.configuredGET(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(expectedParams())
      expect(result[2]).toEqual(expectedRequestInit())
    })

    it('calls configuredDELETE with expected args', async () => {
      const result = await datasource.configuredDELETE(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(expectedParams())
      expect(result[2]).toEqual(expectedRequestInit())
    })

    it('calls configuredPOST with expected args', async () => {
      const result = await datasource.configuredPOST(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(expectedBody())
      expect(result[2]).toEqual(expectedRequestInit())
    })

    it('calls configuredPUT with expected args', async () => {
      const result = await datasource.configuredPUT(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(expectedBody())
      expect(result[2]).toEqual(expectedRequestInit())
    })

    it('calls configuredPATCH with expected args', async () => {
      const result = await datasource.configuredPATCH(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(expectedBody())
      expect(result[2]).toEqual(expectedRequestInit())
    })
  })

  describe('datasource without search params', () => {
    const datasource = new TestDataSourceWithoutSearchParams()

    it('calls configuredGET with expected args', async () => {
      const result = await datasource.configuredGET(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(new URLSearchParams())
      expect(result[2]).toEqual(expectedRequestInit())
    })

    it('calls configuredDELETE with expected args', async () => {
      const result = await datasource.configuredDELETE(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(new URLSearchParams())
      expect(result[2]).toEqual(expectedRequestInit())
    })

    it('calls configuredPOST with expected args', async () => {
      const result = await datasource.configuredPOST(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(expectedBody())
      expect(result[2]).toEqual(expectedRequestInit())
    })

    it('calls configuredPUT with expected args', async () => {
      const result = await datasource.configuredPUT(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(expectedBody())
      expect(result[2]).toEqual(expectedRequestInit())
    })

    it('calls configuredPATCH with expected args', async () => {
      const result = await datasource.configuredPATCH(args)
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toEqual(expectedBody())
      expect(result[2]).toEqual(expectedRequestInit())
    })
  })
})