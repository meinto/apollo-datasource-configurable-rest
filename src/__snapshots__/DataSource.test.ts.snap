// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigurableRESTDataSource datasource with search params calls configuredDELETE with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument"`;

exports[`ConfigurableRESTDataSource datasource with search params calls configuredGET with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument"`;

exports[`ConfigurableRESTDataSource datasource with search params calls configuredPATCH with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument?param1=mock-arg-1&param2=22&param3=%7B%22param31%22%3A%22mock-arg-2%22%2C%22param32%22%3A33%7D"`;

exports[`ConfigurableRESTDataSource datasource with search params calls configuredPOST with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument?param1=mock-arg-1&param2=22&param3=%7B%22param31%22%3A%22mock-arg-2%22%2C%22param32%22%3A33%7D"`;

exports[`ConfigurableRESTDataSource datasource with search params calls configuredPUT with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument?param1=mock-arg-1&param2=22&param3=%7B%22param31%22%3A%22mock-arg-2%22%2C%22param32%22%3A33%7D"`;

exports[`ConfigurableRESTDataSource datasource with search params removes search params and headers when the corresponding arguments are not available 1`] = `"https://test.url/$arg1/mock-arg-2/default-argument"`;

exports[`ConfigurableRESTDataSource datasource with search params removes search params when the corresponding arguments are not available 1`] = `"https://test.url/$arg1/mock-arg-2/default-argument"`;

exports[`ConfigurableRESTDataSource datasource without search params calls configuredDELETE with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument"`;

exports[`ConfigurableRESTDataSource datasource without search params calls configuredGET with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument"`;

exports[`ConfigurableRESTDataSource datasource without search params calls configuredPATCH with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument"`;

exports[`ConfigurableRESTDataSource datasource without search params calls configuredPOST with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument"`;

exports[`ConfigurableRESTDataSource datasource without search params calls configuredPUT with expected args 1`] = `"https://test.url/mock-arg-1/mock-arg-2/default-argument"`;
